pkg/fqdn/regexpmap/regexp_map.go

```diff
@@ -0,0 +1,136 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package regexpmap
+
+import (
+	"regexp"
+	"sort"
+)
+
+// lookupValueSet is a utility type. It is intended as a set of strings
+// inserted into a RegexpMap
+type lookupValueSet map[string]struct{}
+
+// RegexpMap is a map-like type that allows lookups to match regexp keys. These
+// keys are managed internally as strings, and are uniqued by this
+// representation (and not recompiled on repeat inserts).
+// Stored values are strings and are returned as-is, with the exception that
+// repeat inserts of the same value are de-duped.
+// Note: RegexpMap is not thread-safe and managing concurrent access is the
+// responsibility of the callers.
+type RegexpMap struct {
+	// lookups maps the original key string to the values added with it in .Add.
+	// These values are returned when a "key" (the matching regex in rules)
+	// matches.
+	lookups map[string]lookupValueSet
+	// rules maps the original key strong to the compiled regex for it. A .Lookup
+	// call will iterate over all the values in this map.
+	rules map[string]*regexp.Regexp
+}
+
+// NewRegexpMap returns an initialized RegexpMap
+func NewRegexpMap() *RegexpMap {
+	return &RegexpMap{
+		lookups: make(map[string]lookupValueSet),
+		rules:   make(map[string]*regexp.Regexp),
+	}
+}
+
+// Add associates lookupValue as a return value for reStr on lookup
+// Repeated calls with the same reStr will not recompile the regexp, but will
+// store all lookupValues in a set
+func (m *RegexpMap) Add(reStr string, lookupValue string) error {
+	// if this is the first Add of reStr, compile the regexp and setup the
+	// lookupValue set
+	if _, exists := m.rules[reStr]; !exists {
+		rule, err := regexp.Compile(reStr)
+		if err != nil {
+			return err
+		}
+		m.rules[reStr] = rule
+		m.lookups[reStr] = lookupValueSet{}
+	}
+
+	// add the lookupValue to the set for reStr
+	m.lookups[reStr][lookupValue] = struct{}{}
+
+	return nil
+}
+
+// LookupValues returns all lookupValues, inserted via Add, where the reStr
+// matches lookupKey
+func (m *RegexpMap) LookupValues(lookupKey string) (lookupValues []string) {
+	for reStr, rule := range m.rules {
+		if !rule.MatchString(lookupKey) {
+			continue
+		}
+		for lookupValue := range m.lookups[reStr] {
+			lookupValues = append(lookupValues, lookupValue)
+		}
+	}
+	return keepUniqueStrings(lookupValues)
+}
+
+// LookupContainsValue returns true if any reStr in lookups, inserted via Add,
+// matches lookupKey AND has a lookupValue, inserted via the same Add, that
+// matches expectedValue.
+func (m *RegexpMap) LookupContainsValue(lookupKey, expectedValue string) (found bool) {
+	for reStr, rule := range m.rules {
+		if !rule.MatchString(lookupKey) {
+			continue
+		}
+
+		// The values are stored as a set, so a simple map lookup works.
+		if _, found := m.lookups[reStr][expectedValue]; found {
```

Comment:
A good question! Apparently not! https://play.golang.org/p/4pXDWA9gvgg

I guess it returns early if the first map lookup fails. I'll leave a comment since it's not clear at all.